plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBucketing.java

```diff
@@ -0,0 +1,174 @@
+        assertBucketEquals("string", "\uD843\uDFFC\uD843\uDFFD\uD843\uDFFE\uD843\uDFFF"); // 4 code points: 20FFC - 20FFF. 4-byte UTF-8 sequences in Supplementary Plane 2
+
+        assertBucketEquals("decimal(1, 1)", null);
+        assertBucketEquals("decimal(1, 1)", new BigDecimal(1));
+        assertBucketEquals("decimal(1, 1)", new BigDecimal(1));
+        assertBucketEquals("decimal(18, 10)", new BigDecimal(0));
+        assertBucketEquals("decimal(38, 10)", new BigDecimal("999999.9999999999"));
+        assertBucketEquals("decimal(38, 0)", new BigDecimal("99999999999999999999999999999999999999"));
+        assertBucketEquals("decimal(38, 10)", new BigDecimal("9999999999999999999999999999.9999999999"));
+
+        assertBucketEquals("binary", null);
+        assertBucketEquals("binary", ByteBuffer.wrap(new byte[]{}));
+        assertBucketEquals("binary", ByteBuffer.wrap("hello trino".getBytes(StandardCharsets.UTF_8)));
+        assertBucketEquals("binary", ByteBuffer.wrap("\uD843\uDFFC\uD843\uDFFD\uD843\uDFFE\uD843\uDFFF".getBytes(StandardCharsets.UTF_16)));
+    }
+
+    @Test(dataProvider = "unsupportedBucketingTypes")
+    public void testUnsupportedTypes(Type type)
+    {
+        assertThatThrownBy(() -> computeIcebergBucket(type, null, 1))
+                .hasMessage(format("Cannot bucket by type: %s", type));
+
+        assertThatThrownBy(() -> computeTrinoBucket(type, null, 1))
+                .hasMessage(format("Unsupported type for 'bucket': %s", toTrinoType(type, TYPE_MANAGER)));
+    }
+
+    @Test(dataProvider = "unsupportedTrinoBucketingTypes")
+    public void testUnsupportedTrinoTypes(Type type)
+    {
+        assertThatThrownBy(() -> computeTrinoBucket(type, null, 1))
+                .hasMessage(format("Cannot convert from Iceberg type '%s' (%s) to Trino type", type, type.typeId()));
+
+        assertNull(computeIcebergBucket(type, null, 1));
+    }
+
+    @DataProvider
+    public Object[][] unsupportedBucketingTypes()
+    {
+        return new Object[][] {
+            {BooleanType.get()},
+            {FloatType.get()},
+            {DoubleType.get()}};
+    }
+
+    @DataProvider
+    public Object[][] unsupportedTrinoBucketingTypes()
+    {
+        return new Object[][] {
+                {UUIDType.get()}};
+    }
+
+    private void assertBucketEquals(String icebergTypeName, Object value)
+    {
+        Type icebergType = Types.fromPrimitiveString(icebergTypeName);
+        assertBucketNumberEquals(icebergType, value, 2);
+        assertBucketNumberEquals(icebergType, value, 7);
+        assertBucketNumberEquals(icebergType, value, 31);
+        assertBucketNumberEquals(icebergType, value, 32);
+        assertBucketNumberEquals(icebergType, value, 100);
+        assertBucketNumberEquals(icebergType, value, 10000);
+        assertBucketNumberEquals(icebergType, value, Integer.MAX_VALUE);
```

Comment:
let's have this one first; assuming there is `%` later, this one should catch all problems

also, MAX_VALUE is 2^31-1, not a power of two, let's have 2^30 case as maximal power of two
and have some nerdy large prime number case too

Reply:
2^31 - 1 is a prime number